Record and report why the structure diverged for logging and validation

linabutler opened the pull request:
This PR refactors how we resolve diverging structure, and adds a way to report those problems. The diff is pretty big; it's easier to follow the individual commits, which I've tried to break up into reviewable chunks.

The reports are most useful for logging, so we can see why Dogear decided something diverged. Right now, there's no way to tell without examining the DB before a sync, or looking at the server records in About Sync.

We could even turn on debug (trace has PII concerns) and success logging for _just this sync_ if we detect problems. That way, when folks notice issues, they can hopefully upload more complete logs...we wouldn't need to ask them to try to reproduce. We could also explore automated log collection; maybe in Rust Places via Sentry? Though, again, we'd need to make sure there's no PII.

Also, we could use this problem data to report validation errors, like we do on Desktop! It doesn't report non-structure differences...but, since we already have the complete structure, we're already effectively validating before each sync, anyway.

(I think an independent validator in About Sync is still useful—what if Dogear's validator has bugs, or there's a bug in how we fetch or store the tree structure in the DB? Who watches the watcher? 😆 But this might be a better approach for bookmark validation on mobile).

I've also tried to add comments where they make sense, because I don't want Dogear to be write-only. :joy: Let me know if there's anything unclear, or if there are obvious comments that I can remove!

linabutler committed:
Factor out resolving possible parents into `PossibleParents`.

The logic for breaking the tie between conflicting parents is hard to
digest when it's inlined into `resolve`. Instead of using `min_by()`,
we can iterate over a sequence of `PossibleParent`s that implement
the logic as part of `Ord::cmp`.

This commit also adds more comments that explain how we pick the
parent.

linabutler committed:
Factor out parent resolution into `ResolveParent`.

This builds on the last commit. The algorithm for resolving parents
is hard to follow when it's inlined into `Builder::resolve`, isn't
as efficient as it could be for roots, isn't well-commented, and
tolerates unreachable cases.

This commit factors that logic out into a separate `ResolveParent`
type. It simplifies how we handle complete and partial parents,
unrolls the nested iterator chains in `Builder::into_tree`,
tightens up some cases, and adds more docs.

linabutler committed:
Record and report problems with tree structure.

This commit records problems that Dogear notices when building a
tree, and outputs a summary when the tree is printed. It's useful for
logging, so we can see why the structure diverged.

We can also use it to emit validation events, like the Desktop Sync
ping, if we decide to in the future.

linabutler commented on src/tree.rs:
```diff
@@ -827,6 +1005,87 @@ enum Divergence {
+    MissingParentGuid(Guid),
+
+    /// The item's `parentid` exists, but isn't a folder.
+    NonFolderParentGuid(Guid),
```
This is like `childrenOnNonFolder`, but emitted for every child.

linabutler commented on src/tree.rs:
```diff
@@ -827,6 +1005,87 @@ enum Divergence {
+pub enum Problem {
+    /// The item doesn't have a `parentid`, and isn't mentioned in any
+    /// folder's `children`.
+    Orphan,
```
I'm using "orphan" to mean "doesn't have a parent at all, we have no idea where to put it", not "has a `parentid` that doesn't exist, so we can't put it there"—but maybe that distinction isn't useful?

linabutler commented on src/tree.rs:
```diff
@@ -827,6 +1005,87 @@ enum Divergence {
+
+    /// The item is a user content root (menu, mobile, toolbar, or unfiled),
+    /// but its parent isn't the Places root.
+    MisparentedRoot(Guid),
```
I guess this is most like `badClientRoots`, but for the server.

linabutler commented on src/tree.rs:
```diff
@@ -827,6 +1005,87 @@ enum Divergence {
+    MisparentedRoot(Guid),
+
+    /// Multiple folders reference the item in their `children`.
+    MultipleParents(Guid),
```
This is `multipleParents`.

linabutler commented on src/tree.rs:
```diff
@@ -827,6 +1005,87 @@ enum Divergence {
+    MultipleParents(Guid),
+
+    /// The item's `parentid` doesn't match that folder's `children`.
+    ParentChildDisagreement(Guid),
```
This is `parentChildMismatches`.

linabutler commented on src/tree.rs:
```diff
@@ -827,6 +1005,87 @@ enum Divergence {
+    ParentChildDisagreement(Guid),
+
+    /// The item's `parentid` doesn't exist.
+    MissingParentGuid(Guid),
```
This is like `orphans` and `deletedParents`—we don't record deleted items separately from nonexistent ones, but maybe we should?

mhammond replied:
Maybe just `MissingParent` - the `Guid` on the tail made me thing it meant the *guid* was missing in the child, but IIUC, this means the item identified by the guid doesn't exist (and that if the guid was missing it would just be `Orphan`)? Or maybe I'm missing something?n

linabutler commented on src/tree.rs:
```diff
@@ -827,6 +1005,87 @@ enum Divergence {
+
+    /// The item is mentioned in a folder's `children`, but doesn't exist or is
+    /// deleted.
+    MissingChild(Guid),
```
This is `missingChildren` and `deletedChildren`.

linabutler reviewed (commented):
Here are the Sync ping equivalents, more or less. I think all of them would be counted in `structuralDifferences`, too.

thomcc commented on src/tree.rs:
```diff
@@ -769,19 +640,326 @@ enum BuilderParentBy {
+struct ResolveParent<'a> {
+    builder: &'a Builder,
+    entry: &'a BuilderEntry,
+    problems: &'a mut Problems,
```
A little surprised that this having the same lifetime as the non-mut things doesn't lead to problems, but I guess our use is limited.

thomcc reviewed (approved):
Looks fine. Good to see dogear growing more complete bookmark validation!

mhammond reviewed (approved):
This looks great!

linabutler committed:
Improve problem summaries.

* Split `MultipleParents`, `{Missing, NonFolder}ParentGuid`, and
  `ParentChildDisagreement` into `DivergedParent::By{Guid,
  ParentGuid}`, and add `DivergedParentGuid` to distinguish
  between folders, non-folders, and nonexistent parents. This matches
  how the builder stores parents internally, and makes for more
  accurate messages.
* Use named fields in `Problem` variants for clarity.
* Return more readable summary strings.

linabutler committed:
Add more struct and method docs.

linabutler committed:
Only allow `ByParentGuid` overrides for reparented roots.
